### PR TITLE
state(transformation-helpers): slice & extract helpers

### PR DESCRIPTION
**Pull request by Karnaukhov-kh**
**UPD:** After a great suggestion from @hoebbelsB I've renamed object version of `extract` to `slice` and added `extract` as an array helper.

## Idea

We have `selectSlice` for state selections, but we have nothing similar that can be used with `connect`. 

# Slice

## Description
- Accepts an object of type `T` and single key or array of keys (`K extends keyof T`).
- Constructs new object based on provided keys.

## Behavior
- If object is not provided or object is not an object - `undefined` returned.
- If keys are not provided - `undefined` returned.
- If none of the provided keys found or none of the provided keys are valid keys (string, number) - `undefined` returned.
- If some of the keys not found or some of the keys are not valid keys (string, number) - a new object based on valid keys returned.
- If object is object and all provided keys found - a new object based on provided keys returned.

## Example

```typescript
 export class AnimalsListComponent {

    constructor(private state: RxState<ComponentState>, private api: ApiService) {
      state.connect(
        'animals'
        this.api.getAnimals(),
        (state, animals) => {
            return animals.map(animal => slice(animal, ['id', 'name']));
        }
      );
    }
 }
```

## Naming

We talked with @hoebbelsB about the name, but I don't remember the outcome 😄. Believe we talked about `extract` or `slice`. Don't have a strong opinion about this, so we can easily switch from `extract` to `slice`.

# Extract

## Description
- Accepts an array of objects of type T and single key or array of keys (K extends keyof T).
- Constructs new array of objects based on provided keys.

## Behavior
- If array is not provided or array is not an array - `undefined` returned.
- If keys are not provided - `undefined` returned.
- If none of the provided keys found or none of the provided keys are valid keys (string, number) - `undefined` returned.
- `extract([{ id: 1, type: 'cat', name: 'Emma' }, { id: 2, type: 'dog' }], ['name', 'id', 'type'])` > `[{ id: 1, type: 'cat', name: 'Emma' }, { id: 2, type: 'dog', name: undefined }]; // dog has no name`
- `extract([{id: 1, type: 'cat', name: 'Emma'}, 1, 'string'] as any, ['id', 'name'] as any)` > `[{ id: 1,  name: 'Emma' }, { id: undefined, name: undefined }, { id: undefined, name: undefined }]; // cat had properties but they are not found in 1 and 'string'`
- If array is array and all keys found - new array with objects constructed based on keys returned.

## Example

```typescript
 export class AnimalsListComponent {

    constructor(private state: RxState<ComponentState>, private api: ApiService) {
      state.connect(
        'animals'
        this.api.getAnimals(),
        (state, animals) =>  extract(animals, ['id', 'name']))
      );
    }
 }
```

### CODE REVIEW

**Review by hoebbelsB — APPROVED** [review]
thx @Karnaukhov-kh for the contribution! that's a great addition to the transformation helpers.

I was just wondering if it wouldn't be great if we had extract for arrays as well which does the `array.map` part for the user. Your example would look like so:

```ts
state.connect(
  'animals',
  api.getAnimals().pipe(map(animals => extract(animals, ['id', 'name'])))
);
```

**Comment by Karnaukhov-kh** [comment]
Hi @hoebbelsB. As for me, this is a great idea! What do you think about renaming this one to `slice` and introduce `extract` for arrays?

File: libs/state/docs/api/transformation-helpers/object/extract.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+## extract
+
+Accepts an object of type T and single key or array of keys (K extends keyof T).
+Constructs new object based on provided keys.
```
[the remark this answers is not in the document]
Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/object/extract.md:4]: thx 👍

**Comment by hoebbelsB** [comment]
@Karnaukhov-kh this is a really good one :+1:

**Comment by BioPhoton** [comment]
Regarding naming:
Isn't that a "multi pluck" thing?

**Comment by Karnaukhov-kh** [comment]
> Regarding naming:
> Isn't that a "multi pluck" thing?

Kinda. `pluck` allows you to select the field and its children, this one takes multiple properties from the same level just as `selectSlice`

**Comment by Karnaukhov-kh** [comment]
Hi. Renamed `extract` to `slice` and added `extract` for arrays. Updated PR name & first comment according to these changes.

**Review by hoebbelsB — APPROVED** [review]
great job! I have just made some adjustments to the docs. Are they generated on their own after merging btw?

File: libs/state/docs/api/transformation-helpers/object/slice.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-## extract
+## slice
 
 Accepts an object of type T and a single key or an array of keys (`K extends keyof T`).
 Constructs new object based on provided keys.
```
[the remark this answers is not in the document]
Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/object/slice.md]: Hi, thanks. Not sure what to do with the last one. Not sure what you mean by "generate on their own". All docs are made manually. Transformation helpers web page will include `slice.md` & `extract.md` content after merge. I probably should comment it out before release and uncomment after :).

**Comment by BioPhoton** [comment]
Hy @Karnaukhov-kh 
any thing missing here?

File: libs/state/docs/api/transformation-helpers/edge-cases-overview.md [hunk 3]
```diff
@@ -15,11 +15,11 @@ insert(nonArray as any, items) > items;
 ### update
 
 ```typescript
-update(null as any, items) > items;
+update(null as any, items) > null;
 update(items, null as any) > items;
 update(null as any, null as any) > null;
 update(undefined as any, undefined as any) > undefined;
-update(nonArray as any, items) > items;
+update(nonArray as any, items) > nonArray;
```
Comment (hoebbelsB) [inline libs/state/docs/api/transformation-helpers/edge-cases-overview.md:22]: aren't
`update(null as any, items) > null;` &
`update(nonArray as any, items) > nonArray;` 
the same thing?

File: docs/web/state/api/transformation-helpers.md [hunk 4]
```diff
@@ -24,6 +24,10 @@ You can use the transformation helpers without RxState but expect the behavior t
 
 [update.md](https://raw.githubusercontent.com/rx-angular/rx-angular/master/libs/state/docs/api/transformation-helpers/array/update.md ':include')
 
+---
+
+[extract.md](https://raw.githubusercontent.com/rx-angular/rx-angular/master/libs/state/docs/api/transformation-helpers/array/extract.md ':include')
```
Comment (hoebbelsB) [inline docs/web/state/api/transformation-helpers.md:29]: are those links actually working?

File: libs/state/docs/api/transformation-helpers/edge-cases-overview.md (hunk 3 above)
Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/edge-cases-overview.md:22]: True

File: docs/web/state/api/transformation-helpers.md (hunk 4 above)
Comment (Karnaukhov-kh) [reply docs/web/state/api/transformation-helpers.md:29]: They will after merge. :)

File: libs/state/docs/api/transformation-helpers/edge-cases-overview.md (hunk 3 above)
Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/edge-cases-overview.md:22]: Fixed

**Comment by hoebbelsB** [comment]
LGTM. @BioPhoton u want to take a second look?

**Pull request merged**